x/config/loader.go

```diff
@@ -0,0 +1,282 @@
+// THE SOFTWARE.
+
+package config
+
+import (
+	"fmt"
+	"reflect"
+
+	"go.uber.org/yarpc/internal/decode"
+
+	"gopkg.in/yaml.v2"
+)
+
+type transportSpec struct {
+	name                     string
+	transportConfigType      reflect.Type
+	inboundConfigType        reflect.Type
+	unaryOutboundConfigType  reflect.Type
+	onewayOutboundConfigType reflect.Type
+	unaryOutboundPresets     map[string]reflect.Type
+	onewayOutboundPresets    map[string]reflect.Type
+}
+
+func (s *transportSpec) supportsUnaryOutbound() bool {
+	return s.unaryOutboundConfigType != nil
+}
+
+func (s *transportSpec) supportsOnewayOutbound() bool {
+	return s.onewayOutboundConfigType != nil
+}
+
+func (s *transportSpec) transportBuilder(attrs attributeMap) (TransportBuilder, error) {
+	result := reflect.New(s.transportConfigType).Interface()
+	if err := attrs.Decode(result); err != nil {
+		return nil, fmt.Errorf("failed to decode configuration for transport %q: %v", s.name, err)
+	}
+	return result.(TransportBuilder), nil
+}
+
+func (s *transportSpec) inboundBuilder(attrs attributeMap) (InboundBuilder, error) {
+	if s.inboundConfigType == nil {
+		return nil, fmt.Errorf("transport %q does not define an inbound", s.name)
+	}
+
+	result := reflect.New(s.inboundConfigType).Interface()
+	if err := attrs.Decode(result); err != nil {
+		return nil, fmt.Errorf("failed to decode configuration for inbound %q: %v", s.name, err)
+	}
+	return result.(InboundBuilder), nil
+}
+
+func (s *transportSpec) unaryOutboundBuilder(preset string, attrs attributeMap) (UnaryOutboundBuilder, error) {
+	typ := s.unaryOutboundConfigType
+	if typ == nil {
+		return nil, fmt.Errorf("transport %q does not support unary outbounds", s.name)
+	}
+
+	if preset != "" {
+		var ok bool
+		typ, ok = s.unaryOutboundPresets[preset]
+		if !ok {
+			return nil, fmt.Errorf("unknown preset %q for unary outbound %q", preset, s.name)
+		}
+	}
+
+	result := reflect.New(typ).Interface()
+	if err := attrs.Decode(result); err != nil {
+		return nil, fmt.Errorf("failed to decode configuration for unary outbound %q: %v", s.name, err)
+	}
+	return result.(UnaryOutboundBuilder), nil
+}
+
+func (s *transportSpec) onewayOutboundBuilder(preset string, attrs attributeMap) (OnewayOutboundBuilder, error) {
+	typ := s.onewayOutboundConfigType
+	if typ == nil {
+		return nil, fmt.Errorf("transport %q does not support oneway outbounds", s.name)
+	}
+
+	if preset != "" {
+		var ok bool
+		typ, ok = s.onewayOutboundPresets[preset]
+		if !ok {
+			return nil, fmt.Errorf("unknown preset %q for oneway outbound %q", preset, s.name)
+		}
+	}
+
+	result := reflect.New(typ).Interface()
+	if err := attrs.Decode(result); err != nil {
+		return nil, fmt.Errorf("failed to decode configuration for oneway outbound %q: %v", s.name, err)
+	}
+	return result.(OnewayOutboundBuilder), nil
+}
+
+// Loader TODO
+type Loader struct {
+	knownTransports map[string]transportSpec
+}
+
+// NewLoader TODO
+func NewLoader() *Loader {
```

Comment:
Aaah, looks like `RegisterTransport()` is in another file, suggest moving that close to here:

1. NewLoader
1. RegisterLoader
1. Load
1. LoadYAML